bench/benchmarks_view.js

```diff
@@ -423,41 +432,103 @@ for (const name in window.mapboxglBenchmarks) {
     if (filter && name !== filter)
         continue;
 
-    const benchmark = { name, versions: [] };
-    benchmarks.push(benchmark);
-
-    for (const ver in window.mapboxglBenchmarks[name]) {
-        const version = {
-            name: ver,
-            status: 'waiting',
-            logs: [],
-            samples: [],
-            summary: {}
-        };
-
-        benchmark.versions.push(version);
-
-        promise = promise.then(() => {
-            version.status = 'running';
-            update();
-
-            return window.mapboxglBenchmarks[name][ver].run()
-                .then(measurements => {
-                    // scale measurements down by iteration count, so that
-                    // they represent (average) time for a single iteration
-                    const samples = measurements.map(({time, iterations}) => time / iterations);
-                    version.status = 'ended';
-                    version.samples = samples;
-                    version.summary = summaryStatistics(samples);
-                    version.regression = regression(measurements);
-                    update();
-                })
-                .catch(error => {
-                    version.status = 'errored';
-                    version.error = error;
+    if (isStyleBench && name === 'Layout' || name === 'Paint') {
+        // create a new test in the requested benchmark suite for each location
+        // this benchmarks array is distinct from window.mapboxglBenchmarks and is used to create and update the UI
+        locations.forEach(location => {
+            benchmarks.push({benchmark: {
+                location,
+                name
+            }});
+        });
+
+        for (const loc in window.mapboxglBenchmarks[name]) {
+            // we have to add the versions array here
+            // otherwise, we end up duplicating tests
+            benchmarks.forEach(bench => {
+                if (bench.hasOwnProperty('benchmark')) {
+                    bench.benchmark.versions = [];
+                }
+            });
+
+            for (const style in window.mapboxglBenchmarks[name][loc]) {
+                // push a test object for each style
+                benchmarks.forEach(bench => {
+                    if (bench.hasOwnProperty('benchmark')) {
+                        bench.benchmark.versions.push({
+                            name: style,
+                            status: 'waiting',
+                            logs: [],
+                            samples: [],
+                            style: {},
+                            summary: {}
+                        });
+                    }
+                });
+
+                promise = promise.then(() => {
```

Comment:
This seems to be in tension with @jfirebaugh's suggestion above.

> It looks like there's more divergent branches than shared code in some of these files, so rather than trying to share all the same files between the two benchmarking modes, what about an approach where we use different paths/index.html/benchmarks.js files?

The biggest changes are this section and the `register` function in `bench/benchmarks.js`. For the individual benchmark tests, the simplest thing from a code standpoint might be to just standardize on the `style_locations` as input. We don't have to run the tests individually since that takes a long time, but we could still use the `style_locations` as opposed to a simple array of zoom numbers with a hardcoded center all averaged together into a single test. You can see this in action in the `QueryBox` and `QueryPoint` benches because that's exactly what they do in a style test. This might give us better real-world tests in version benchmarking while simplifying the code base some.

As for this section, is it better to attempt to factor out the similar code here or break it apart the way John suggests?

Reply:
:thinking: Yeah there is some tension, but maybe not that much? To me it makes sense that specifying the set of benchmarks (e.g. `benchmarks.js`) would have different implementations, but rendering a benchmark (e.g. `benchmarks_view.js`) would have a shared/generalized implementation.

Reply:
What do you think about standardizing on the locations/tiles in `style_locations.js`? I updated the handful of tests that rely on iterating over a few different zoom levels to use the ones in `style_locations` and it all still works. The code is simpler as well. The `setup` method in `Paint`, for example, went from 25 to 11 lines of code. The tradeoff is that the tests take longer now, which you'd expect from benching 15 maps as opposed to 4 or 5. It's most dramatic in `Layout`.

![screen shot 2018-07-31 at 3 25 55 pm](https://user-images.githubusercontent.com/4523080/43490724-a6899968-94d6-11e8-81be-f12c541621af.png)

Reply:
I like the idea of using a standardized code pathway/definition format for specifying multiple locations. I think we should probably have separate location definitions for style benchmarking vs. the regular benchmarks because:
 - We'll probably want to change the style locations more often
 - In style benchmarking we'll typically care about more locations and be willing to spend more time testing them
 - Changing the location set isn't a big deal when you're comparing multiple styles with the same code version, but since it's baked into the benchmark, changing the location set is a big deal for comparing between code versions (as shown in your screenshot above: it breaks comparisons for Layout whenever different versions have different location sets).